### ISSUE TITLE
Test runs are slower than IntelliJ

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#749]
Reported by @kentcb in Gitter:

I have the latest dart-code version and am trying to get unit tests running. I added this configuration:

```js
        {
            "name": "Unit Tests",
            "type": "dart",
            "request": "launch",
            "program": "${workspaceFolder}/lib/tests/unit_tests.dart"
        },
```

It actually runs, and I see test output in the debug console. However, there are several problems I'd like some clarity on:

- ~~Breakpoints are not hit~~ #748
- Test execution is very slow compared to IntelliJ's unit testing. It seems to be creating an APK and launching the tests on the device/emulator, whereas they seem to run without this in IntelliJ.
- ~~Once tests complete, the process does not exit. I have to hit stop (Shift+F5) to terminate.~~ #750

**Comment by DanTup** [comment]
I ran the `locale_test.dart` file from the Flutter Stocks app and it took 6 seconds. The logs show that 4 seconds were waiting for test process to start:

```text
[10:05:23]: Spawning /Users/dantup/Dev/Google/flutter/bin/flutter with args ["test","--machine","--start-paused","test/locale_test.dart"]
[10:05:23]: ..  in /Users/dantup/Dev/Google/flutter/examples/stocks
[10:05:23]: ..  in {"FLUTTER_HOST":"VSCode"}
[10:05:24]: <== {"protocolVersion":"0.1.0","runnerVersion":"0.12.32+2","type":"start","time":0}
[10:05:24]: <== {"count":1,"type":"allSuites","time":0}
[10:05:24]: <== {"suite":{"id":0,"platform":"vm","path":"/Users/dantup/Dev/Google/flutter/examples/stocks/test/locale_test.dart"},"type":"suite","time":0}
[10:05:24]: <== {"test":{"id":1,"name":"loading /Users/dantup/Dev/Google/flutter/examples/stocks/test/locale_test.dart","suiteID":0,"groupIDs":[],"metadata":{"skip":false,"skipReason":null},"line":null,"column":null,"url":null},"type":"testStart","time":0}
[10:05:28]: <== [{"event":"test.startedProcess","params":{"observatoryUri":"http://127.0.0.1:55053/"}}]
[10:05:28]: <== {"testID":1,"result":"success","skipped":false,"hidden":true,"type":"testDone","time":4460}
[10:05:28]: <== {"group":{"id":2,"suiteID":0,"parentID":null,"name":null,"metadata":{"skip":false,"skipReason":null},"testCount":1,"line":null,"column":null,"url":null},"type":"group","time":4464}
[10:05:28]: <== {"group":{"id":3,"suiteID":0,"parentID":2,"name":"-","metadata":{"skip":false,"skipReason":null},"testCount":1,"line":58,"column":16,"url":"package:flutter_test/src/widget_tester.dart"},"type":"group","time":4464}
[10:05:28]: <== {"test":{"id":4,"name":"- Changing locale","suiteID":0,"groupIDs":[2,3],"metadata":{"skip":false,"skipReason":null},"line":59,"column":18,"url":"package:flutter_test/src/widget_tester.dart"},"type":"testStart","time":4465}
[10:05:29]: <== {"testID":4,"result":"success","skipped":false,"hidden":false,"type":"testDone","time":5104}
[10:05:29]: <== {"success":true,"type":"done","time":5128}
```

Are you seeing significantly slower, or is IntelliJ much faster?

I tried running without the debugger (`Ctrl`+`F5`) to see if it was that slowing things down, but it was exactly the same:

```text
[10:07:09]: Spawning /Users/dantup/Dev/Google/flutter/bin/flutter with args ["test","--machine","test/locale_test.dart"]
[10:07:09]: ..  in /Users/dantup/Dev/Google/flutter/examples/stocks
[10:07:09]: ..  in {"FLUTTER_HOST":"VSCode"}
[10:07:10]: <== {"protocolVersion":"0.1.0","runnerVersion":"0.12.32+2","type":"start","time":0}
[10:07:10]: <== {"count":1,"type":"allSuites","time":0}
[10:07:10]: <== {"suite":{"id":0,"platform":"vm","path":"/Users/dantup/Dev/Google/flutter/examples/stocks/test/locale_test.dart"},"type":"suite","time":0}
[10:07:10]: <== {"test":{"id":1,"name":"loading /Users/dantup/Dev/Google/flutter/examples/stocks/test/locale_test.dart","suiteID":0,"groupIDs":[],"metadata":{"skip":false,"skipReason":null},"line":null,"column":null,"url":null},"type":"testStart","time":0}
[10:07:14]: <== {"testID":1,"result":"success","skipped":false,"hidden":true,"type":"testDone","time":4340}
[10:07:14]: <== {"group":{"id":2,"suiteID":0,"parentID":null,"name":null,"metadata":{"skip":false,"skipReason":null},"testCount":1,"line":null,"column":null,"url":null},"type":"group","time":4343}
[10:07:14]: <== {"group":{"id":3,"suiteID":0,"parentID":2,"name":"-","metadata":{"skip":false,"skipReason":null},"testCount":1,"line":58,"column":16,"url":"package:flutter_test/src/widget_tester.dart"},"type":"group","time":4343}
[10:07:14]: <== {"test":{"id":4,"name":"- Changing locale","suiteID":0,"groupIDs":[2,3],"metadata":{"skip":false,"skipReason":null},"line":59,"column":18,"url":"package:flutter_test/src/widget_tester.dart"},"type":"testStart","time":4343}
[10:07:15]: <== {"testID":4,"result":"success","skipped":false,"hidden":false,"type":"testDone","time":4987}
[10:07:15]: <== {"success":true,"type":"done","time":5010}
```

Running from the command line is marginally faster (reports 4 seconds, but I think it was more like 5-6 end-to-end):

```text
dantup-macbookpro:stocks dantup$ flutter test test/locale_test.dart 
00:04 +1: All tests passed!                                           
```

Let me know what the difference you're seeing it. You can enable the above logging by following instructions here:

https://dartcode.org/docs/logging/#flutter-run

**Comment by DanTup** [comment]
@kentbc Please shout if you can still repro this and provide a log.

**Comment by kentcb** [comment]
This is definitely a case of IntelliJ being faster, because IntelliJ is doing something quite different to VS Code...

When I run my tests in IntelliJ, they run without building the APK and without deploying it to a device/emulator. In fact, there is no connection made to the emulator it all. When I run via VSCode (using the above configuration) I see the APK built and deployed, and I see (and hear) all my widget tests run in the emulator. And it is _very_ slow in comparison.

This is what my IntelliJ test configuration looks like, for the record:

![image](https://user-images.githubusercontent.com/1901832/39611576-fbf4ff30-4f9b-11e8-9b8d-c8701b2dd760.png)

I dug into this a bit and found that IntelliJ starts the following process when I execute tests:

"C:\Users\kent\Repository\flutter\bin\cache\dart-sdk\bin\dart.exe"   "C:\Users\kent\Repository\flutter\bin\cache\flutter_tools.snapshot" --no-color test --machine --preview-dart-2 lib\tests\tests.dart

I tried running this from the command line, and it also executes quickly (compared to running with `flutter test`).

Now, I'm _guessing_ it's possible for me to define a configuration in VS Code that starts the above process instead of `flutter test`, but I'm not sure how. I also think it's worth understanding why IntelliJ uses the above instead of `flutter test`, and whether VS Code should follow suit.

**Comment by DanTup** [comment]
Dart Code also should not be running in the emulator or building the SDK :/

The command you posted:

```
"C:\Users\kent\Repository\flutter\bin\cache\dart-sdk\bin\dart.exe"   "C:\Users\kent\Repository\flutter\bin\cache\flutter_tools.snapshot" --no-color test --machine --preview-dart-2 lib\tests\tests.dart
```

Is just running `flutter run`. The `flutter` script just runs `dart (snapshot)`, so this should be the same as doing:

```
"C:\Users\kent\Repository\flutter\bin\flutter" --no-color test --machine --preview-dart-2 lib\tests\tests.dart
```

I think the problem here is that your tests are in an unexpected location (`./lib/tests` instead of `./test`) so we're not detecting them as tests, so they're being run with `flutter run`. Can you try moving them and let me know if that makes a difference?

**Comment by kentcb** [comment]
Hmm, not having much luck here, but I may be doing something wrong still. I added a `test/tests.dart` file, which has a `main` that simply forwards on to my existing tests:
```dart
import 'package:sandvik/tests/tests.dart' as tests;

void main() {
  tests.main();
}
```
I then updated my configuration in VS Code to:
```json
        {
            "name": "Tests",
            "type": "dart",
            "request": "launch",
            "program": "${workspaceFolder}/test/tests.dart"
        },
```
But this still builds the APK and launches in the simulator. I had a look on https://dartcode.org to try and find some guidance on setting up tests, but I can't see anything.

Am I doing it wrong?

**Comment by DanTup** [comment]
I think you may be hitting #888.

Can you try:

1. Changing `${workspaceFolder}` to the actual full path; or
2. Deleting/moving `launch.json` away, opening `test/tests.dart` in the editor and then pressing `F5`?

Let me know if either of those work, and if so I think it's the bug above. Waiting to hear back from MS but might have to put a workaround in for vNext.

**Comment by kentcb** [comment]
Yeah, option 1 totally works. Moreover, breakpoints work in my tests (they don't in IntelliJ - never figured out why).

I'm happy to run with a hard-coded path for now, but agreed it would be great to get this fixed.

I think this finally enables me to start using VSCode for flutter work - thanks Danny!

**Comment by DanTup** [comment]
Great; then this is fixed in `master` by the changes made for #888. Probably there will be a beta release end of next week.
